### PR TITLE
Improve isotropic phase field fracture model

### LINKED ISSUE #10774: Improve isotropic phase field fracture model
## Rationale

The current isotropic phase field fracture model suffers from poor convergence. The main reason are
1. The jacobian calculation is not complete
2. The use of ramp function to enforce irreversibility reduces the smoothness of solution

## Description

To improve the convergence,
1. Correct the jacobian calculation, particularly for the projection tensor
2. Use history variable to replace ramp function

Besides, some extra work
 1. RankTwoTensor: added mixed product IljK
 2. remove PFFracBulkRateBase, PFFracBulkRate, SplitPFFractureBulkRate
 3. remove AllenCahnPFFracture, replaced by PFFractureBulkRate, but still can use AllenCahn Action
 4. remove PFFracBulkRateMaterial, replaced by GenericFunctionMaterial
 5. replace SplitPFFractureBulkReate with HyperElasticPFFractureBulkRate.C since one test in phase_field_fracture_viscoplastic still needs to use the old kernel. Need to move the HyperElastic phase fracture model to this new formulation in the future. 

## Impact

This might affect some tests in Marmot which use phase field fracture model

### PR DESCRIPTION
**Pull request by jiangwen84**
* RankTwoTensor: added mixed product IljK
* removed PFFracBulkRateBase, PFFracBulkRate, SplitPFFractureBulkRate
* removed AllenCahnPFFracture, replaced by PFFractureBulkRate, but still can use AllenCahn Action
* improve jacobian in ComputeIsotropicLinearElasticPFFractureStress
* removed PFFracBulkRateMaterial, replaced by GenericFunctionMaterial
* cleanup and regold tests
* replaced SplitPFFractureBulkReate with HyperElasticPFFractureBulkRate.C since one test in phase_field_fracture_viscoplastic still needs to use the old kernel.

closes  #10774

### CODE REVIEW

**Comment by sapitts** [comment]
@jiangwen84 would you add documentation in the md file for ComputeIsotropicLinearElasticPFFractureStress in tensor mechanics while you are working with that file please?

**Comment by jiangwen84** [comment]
@sapitts I guess you mean ``ComputeLinearElasticPFFractureStress``? I will add that one when I work on that. 

This PR only modifies ``ComputeIsotropicLinearElasticPFFractureStress`` and that one has been documented but I still need to make some changes.

**Comment by sapitts** [comment]
Thanks @jiangwen84, that's the class I meant.

**Comment by tonkmr** [comment]
@jiangwen84, I disagree completely with your choice to bring back a unique kernel that only works for the phase field fracture model. The fracture model is still a phase field method, and so anyone who uses it really needs to understand the basics of phase field. 

There are many reasons why it is very useful to have the fracture model use general phase field kernels. First, as improvements are made to the phase field module, they will automatically improve the fracture model (like the Nonconserved action). Second, it will make it easier to couple the fracture model to other phase field models. Third, I actually think it makes it easier to understand the fracture model. Previously, we got many questions about the model because it was so different than everything else that was in MOOSE. 

Please continue to use the energy for the fracture model in the standard AllenCahn kernel. Then, the anisotropric model would simply be a different energy function. My long term plans are to develop an approach to make the constitutive model plug and play, so that the stress would be calculated by  the constitutive model and then modified by the phase field fracture energy function.

**Comment by jiangwen84** [comment]
@tonkmr I cannot completely agree.  In my opinion, it is valuable to have a separate phase field fracture kernel. 

First, users does not need to specify additional material properties, such as kappa, L. They are only used to match the general ACkernel. In the fracture model,  gc, l and viscosity have real physical meanings but not kappa ( kappa = gc * l) and L ( L= 1/(gc_prop * visco)). Second, PF fracture model do not always need a time derivative kernel. Users can just use one PFFracture Kernel to evolve damage. Third, for phase field fracture model, I could not find the advantage of separating separate fracture energy to two different kernels (ACInterface and ACBulk).

I could get rid of the PFFracture Kernel and just use AC kernels as what you suggest,  but it is hard to convince myself we have to do that.

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C [hunk 1]
```diff
@@ -119,26 +114,75 @@ ComputeIsotropicLinearElasticPFFractureStress::computeQpStress()
     _hist[_qp] = _hist_old[_qp];
 
   // Damage associated with positive component of stress
-  _stress[_qp] =
-      stress0pos * cfactor * ((1.0 - c) * (1.0 - c) * (1 - _kdamage) + _kdamage) - stress0neg;
+  _stress[_qp] = stress0pos * ((1.0 - c) * (1.0 - c) * (1 - _kdamage) + _kdamage) - stress0neg;
 
   // Elastic free energy density
-  _F[_qp] = _hist[_qp] * cfactor * ((1.0 - c) * (1.0 - c) * (1 - _kdamage) + _kdamage) - G0_neg +
+  _F[_qp] = _hist_old[_qp] * ((1.0 - c) * (1.0 - c) * (1 - _kdamage) + _kdamage) - G0_neg +
```
Comment (tonkmr) [inline modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: I'm not sure we want to hard code this to always use the old version of the history variable. Shuaifang has seen some results where it gives strain results using the old version of the history variable. Maybe make it an option of the user to choose to use old or current.

**Comment by tonkmr** [comment]
I have learned by experience that creating a unique kernel for each model becomes very unwieldy very fast. When we first started MARMOT, we would create a new kernel for every model and made things really messy and hard to manage. Rather than create unique kernels just for fracture, we can have kappa and L get set from gc and l in the fracture material or in an additional material.

**Comment by tonkmr** [comment]
I'm also unclear on how you could not have a time derivative kernel. The change in eta is defined by the residual equation because of the time derivative kernel. Without it, it seems like you will lose the physical basis for the evolution.

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C (hunk 1 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: In my opinion, I think we should always use the old value and not the current one since the history variable is defined as max[0-t_n]. If we update this variable during the current iteration, it should not be called 'history' variable anymore. Actually most of paper use staggered approach and they automatically use the old value for this.

The use of the current value also makes convergence bad. One more thing, if we want to use ACKernel, the derivatives of this term respect to displacement cannot be automatically added to off-diagonal Jacobian.

**Comment by jiangwen84** [comment]
No! You can definitely remove the time derivative kernel. The change of eta is driven by the stress change rather than the its rate. We solve the phase field order parameter to  its steady state and it is called rate independent model in Mieche's original paper and used in many other papers as well. 

Actually I do not quite understand the viscosity parameter. It just scales the time step dt. I do not think it can be quantified from either experiment or theory. When we have convergence issue, we are told to use higher value and it essentially make the time step smaller. Since the new formulation has much better convergence, I  would just not use this term at all.

**Comment by jiangwen84** [comment]
@tonkmr I saw your point that the use of less kernels will be easier to maintain the code. I will remove the PF specific kernel and just use AC kernel. I will also add one test without using the time derivative. In addition, I can add the option to use the current history value but set the default value to be false.

**Comment by jiangwen84** [comment]
@dschwen I changed the codes according to Mike's suggestions. So I think now it is ready for you to review the code now? Thanks.

**Comment by dschwen** [comment]
I'm on travel. I'll review this as soon as I have reliable internet.
Probably later today.

**Review by dschwen — DISMISSED** [review]
Looks good to me. This removes a lot of old code. The non-split PFF model makes no sense anymore?

**Comment by jiangwen84** [comment]
I guess so. The non-split PFF version is based on the old formulation and has some convergence issues.

**Comment by jiangwen84** [comment]
The only concern is that we use the stateful material property and I am not sure if the current adaptivity capability supports that or not.

File: modules/phase_field/include/kernels/HyperElasticPFFractureBulkRate.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef HYPERELASTICPFFRACTUREBULKRATE_H
```
[the remark this answers is not in the document]
Comment (tonkmr) [reply modules/phase_field/include/kernels/HyperElasticPFFractureBulkRate.h]: @jiangwen84 can you answer this question for me?

**Review by tonkmr — APPROVED** [review]

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C [hunk 3]
```diff
@@ -119,26 +114,75 @@ ComputeIsotropicLinearElasticPFFractureStress::computeQpStress()
...
+        proj_neg += (Gab + Gba) * 0.5;
+      }
+    }
+  }
 
-  // Estimate Jacobian mult, exact when c = 0 and VERY APPROXIMATE when c = 1
-  _Jacobian_mult[_qp] = _elasticity_tensor[_qp];
+  _Jacobian_mult[_qp] = ((1.0 - c) * (1.0 - c) * (1 - _kdamage) + _kdamage) *
+                            (lambda * dtraceAdA * MathUtils::heavyside(etr) + 2 * mu * proj_pos) +
+                        (lambda * dtraceAdA * MathUtils::heavyside(-etr) + 2 * mu * proj_neg);
```
[the remark this answers is not in the document]
Comment (tonkmr) [reply modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: @jiangwen84 please answer this question. Since the Jacobian is such a hard issue here, I am hoping you have it all figured out ;)

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C [hunk 4]
```diff
@@ -119,26 +114,75 @@ ComputeIsotropicLinearElasticPFFractureStress::computeQpStress()
...
+    Ma.vectorOuterProduct(eigvec.column(i), eigvec.column(i));
+    Mb.vectorOuterProduct(eigvec.column(i), eigvec.column(i));
+    Gab = Ma.outerProduct(Mb);
+    proj_pos += MathUtils::heavyside(eigval[i]) * Gab;
+    proj_neg += (1.0 - MathUtils::heavyside(eigval[i])) * Gab;
+  }
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    for (unsigned int j = 0; j < LIBMESH_DIM && j != i; ++j) // Take care for this part, check it
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: got it!

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: The Jacobian should be correct and you can definitely use Newton. It is close to the optimal rate of convergence. The below shows the NEWTON convergence for a time step when the crack is propagating. 

 0 Nonlinear |R| = 4.545446e+02
      0 Linear |R| = 4.545446e+02
      1 Linear |R| = 2.468107e-13
 1 Nonlinear |R| = 4.888067e-03
      0 Linear |R| = 4.888067e-03
      1 Linear |R| = 4.210976e-17
 2 Nonlinear |R| = 4.291498e-04
      0 Linear |R| = 4.291498e-04
      1 Linear |R| = 1.035821e-18
 3 Nonlinear |R| = 6.833061e-06
      0 Linear |R| = 6.833061e-06
      1 Linear |R| = 1.049054e-20
 4 Nonlinear |R| = 2.992739e-09
      0 Linear |R| = 2.992739e-09
      1 Linear |R| = 9.916736e-24

File: modules/phase_field/include/kernels/HyperElasticPFFractureBulkRate.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/phase_field/include/kernels/HyperElasticPFFractureBulkRate.h]: Good point!

I was not quite familiar with the HyperElastic formation which is implemented here. I wonder if you have any reference. I do not want to make any brute changes without fully understanding. 

I guess I could change HyperElasticPhaseFieldIsoDamage material class following the ComputeIsotropicLinearElasticPFFractureStress and use AC kernels to solve the PF. But I cannot guarantee the Jacobian is correct there.

[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/phase_field/include/kernels/HyperElasticPFFractureBulkRate.h]: @tonkmr Let me work on this and keep you posted.

**Comment by jiangwen84** [comment]
@tonkmr I removed the HyperElastic specific kernel and now use ACKernel. I did not change the Jacobian dstress/dstrain term since this is calculated inside ``FiniteStrainHyperElasticViscoPlastic`` from which ``HyperElasticPhaseFieldIsoDamage`` is derived from. 

@dschwen I moved the calculation of the projection tensor into RankTwoTensor since this basic calculation will be reused for other cases as well.

**Comment by jiangwen84** [comment]
@dschwen Can you help merge the codes? Thanks!

**Comment by jiangwen84** [comment]
I separated the documentation out of this PR and will submit another PR once the new documentation system is complete.

@dschwen Could you want to take a look at it and merge if it does not need any further changes? Thanks!

File: framework/src/utils/RankTwoTensor.C [hunk 5]
```diff
@@ -580,6 +594,56 @@ RankTwoTensor::mixedProductJkIl(const RankTwoTensor & b) const
...
+  // The calculate of projection tensor follows
+  // C. Miehe and M. Lambrecht, Commun. Numer. Meth. Engng 2001; 17:337~353
+
+  // Compute eigenvectors and eigenvalues of mechanical strain
+  RankTwoTensor eigvec;
+  std::vector<Real> eigval(N);
+  (*this).symmetricEigenvaluesEigenvectors(eigval, eigvec);
+
+  // Separate out positive and negative eigen values
+  std::vector<Real> epos(N);
```
Comment (dschwen) [inline framework/src/utils/RankTwoTensor.C]: how about using a `std::array<Real, N>` here?

Comment (jiangwen84) [reply framework/src/utils/RankTwoTensor.C]: sounds good! Done!

**Comment by jiangwen84** [comment]
@dschwen any other comments? Thanks!

File: framework/src/utils/RankTwoTensor.C [hunk 6]
```diff
@@ -580,6 +594,56 @@ RankTwoTensor::mixedProductJkIl(const RankTwoTensor & b) const
...
+RankFourTensor
+RankTwoTensor::positveProjectionEigenDecomposition() const
+{
+  // The calculate of projection tensor follows
+  // C. Miehe and M. Lambrecht, Commun. Numer. Meth. Engng 2001; 17:337~353
+
+  // Compute eigenvectors and eigenvalues of mechanical strain
+  RankTwoTensor eigvec;
+  std::vector<Real> eigval(N);
+  (*this).symmetricEigenvaluesEigenvectors(eigval, eigvec);
```
Comment (dschwen) [inline framework/src/utils/RankTwoTensor.C]: the `(*this).` (which should always be written as `this->` for clarity) is not needed here. It'd be neat fo we could change the vector above to an array as well, but that might require templating or overloading the `symmetricEigenvaluesEigenvectors` member. Let's do those optimizations in a different PR.

**Review by dschwen — DISMISSED** [review]
If you want to change the `(*this).` great, if not no big deal. Looks good to me.

File: framework/src/utils/RankTwoTensor.C (hunk 6 above)
Comment (jiangwen84) [reply framework/src/utils/RankTwoTensor.C]: Sounds great. I will address them in next PR.

**Comment by jiangwen84** [comment]
@dschwen Can you merge the codes? Thanks!

**Review by dschwen — APPROVED** [review]

**Comment by dschwen** [comment]
@tonkmr we went over the discussion and I could not see any unaddressed points. I've merged this in and we can address any further issues in follow-on PRs.

File: modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C]: @tonkmr In case you did not see my reply regarding the jacobian
